procedures: rewrite and expand the "Backup and disaster recovery" sections of the admin guide (#2121)
codenvy/che-docs: modules/administration-guide/partials/con_supported-restic-compatible-backup-servers.adoc

[id="supported-restic-compatible-backup-servers_{context}"]
= Supported restic-compatible backup servers

> In a production environment, {prod-short} uses {prod-short} Operator and restic for backing up and restoring {prod-short} instances from backup snapshots on a configured external backup server. To back up, {prod-short} Operator gathers the data needed for a backup snapshot, and then restic creates and manages the snapshot. To restore, restic retrieves and decrypts the snapshot, and then {prod-short} Operator applies the retrieved data to perform the recovery.

Reviewer: > In a production environment

In every environment actually.

Reviewer: What this paragraph says is true, but `restic` is used under the hood. Admin don't have to use restic at all. In fact, when a backup or restore requested (by chectl or manually be creating a CR - doesn't matter) everything is done by Operator. When reading this I have impression that I have to use restic somewhere in the flow (it is mentioned in every sentence!), but this is no true.
I'd definitely keep the explanation about restic (I think that telling about such thing to admins is useful), but do not bring it front and move to a sort of note somewhere at the end of the page. Maybe even explicitly say that admins don't have to deal with restic directly. Another note can say, that restic might be used to look what is stored in the backup server, for example finding needed snapshot id if it was lost.

Author: Corrected based on your feedback:

`{prod-short} uses {prod-short} Operator and integrated Restic for backing up and restoring {prod-short} instances from backup snapshots on a configured external backup server. To back up, {prod-short} Operator gathers the data needed for a backup snapshot and uses the integrated Restic to create and manage the snapshot. To restore, {prod-short} Operator uses the integrated Restic to retrieve and decrypt the snapshot, and then {prod-short} Operator applies the retrieved data to perform the recovery.`

> {prod-short} can use the following external backup servers that are compatible with restic:

Reviewer: I can set up any kind of the supported backup servers internally in the cluster as well.

Author: I removed "external" from "backup servers". If I understand it correctly, we are not planning to encourage admins to set up backup servers inside the cluster, so I'm not mentioning that possibility (as it's not the expected use case).

> {prod-short} uses {prod-short} Operator and integrated Restic for backing up and restoring {prod-short} instances from backup snapshots on a configured backup server. {prod-short} Operator automates the creation of a Restic backup repository on the backup server. To back up, {prod-short} Operator gathers the data needed for a backup snapshot and uses the integrated Restic to create and manage the snapshot. To restore, {prod-short} Operator uses the integrated Restic to retrieve and decrypt the snapshot, and then {prod-short} Operator applies the retrieved data to perform the recovery.

Reviewer 2: ```suggestion
{prod-short} uses the {prod-short} Operator and integrated Restic to back up and restore {prod-short} instances from backup snapshots on a configured backup server. The {prod-short} Operator automates the creation of a Restic backup repository on the backup server. To back up data, the {prod-short} Operator gathers the data required for a backup snapshot and uses Restic to create and manage the snapshot. To restore data, the {prod-short} Operator uses Restic to retrieve and decrypt the snapshot, and then the {prod-short} Operator applies the retrieved data to perform the recovery.
```
- `Che Operator` needs a definite article `the Che Operator`
- You don't need to keep repeating `integrated Restic` after stating it once in the first sentence.

Author: @bburt-rh Thank you! Fixed.

> SFTP:: See the documentation for the SFTP server solution you plan to use (link:https://www.openssh.com/[OpenSSH.com] or a commercial product) and link:https://restic.readthedocs.io/en/latest/030_preparing_a_new_repo.html#sftp[Restic Docs on SFTP].

Reviewer: > or a commercial product

What do you mean?

Author: It looks like OpenSSH is an open source tool and there are some commercial products based on it.

> SFTP:: See the documentation for the SFTP server solution you plan to use (link:https://www.openssh.com/[OpenSSH] or a derived commercial product) and link:https://restic.readthedocs.io/en/latest/030_preparing_a_new_repo.html#sftp[Restic Docs on SFTP].

Reviewer 2: ```suggestion
SFTP:: See the documentation for the SFTP server solution you plan to use (link:https://www.openssh.com/[OpenSSH] or a derived commercial product) and the link:https://restic.readthedocs.io/en/latest/030_preparing_a_new_repo.html#sftp[Restic Docs on SFTP].
```

Author: Thank you. Fixed.

> SFTP:: See the documentation for the SFTP server solution you plan to use (link:https://www.openssh.com/[OpenSSH] or a derived commercial product) and the link:https://restic.readthedocs.io/en/latest/030_preparing_a_new_repo.html#sftp[Restic Docs on SFTP].

Reviewer: nitpick: I'd like to put it at the end of the list, please, or even better, reverse the order to have rest, s3, sftp.

Author: @mmorhun
I understand that I'm not a typical admin. That said, on another page, we recommend a specific [REST server ](https://github.com/restic/rest-server) with which I ran into the following issues when I was trying to set it up:
1. Neither `restic` nor `rest server` show any error messages if something in the setup is missing or not as expected. So trying this combination of both `restic` nor `rest server` to work together using a remote server turned into frustrating guesswork for me.
2. The README on https://github.com/restic/rest-server is brief and unclear, and I couldn't find online any third-party explanation about the details of how it should be used correctly (without experimenting to make it work "trial and error"). I believe they need to improve their docs.
3. I was setting up the external server from scratch, so I ended up having to research and install some dependencies for the `rest server` that I didn't have to worry about for SFTP.

So I killed a lot of time trying to make the `rest server` work on a remote server, and then in contrast I quickly and easily set up an SFTP server. BTW, SFTP also involves an open source solution.

The page on REST servers is full-size in this PR. It's just that I moved it to the third place for users who have no preference and just pick the first option (like I did). If any users prefer to use `rest server`, they still have everything we expect to give them in the docs.

Reviewer: @max-cx thank you for explaining the reasons. In my case, REST server was the easiest solution, it just worked, where as I had to deal with SSH keys and proper settings for SSH server in case of SFTP (not a problem, but took more time than for REST server).
@tolusha what do you think about this?

> Amazon S3:: See link:https://docs.aws.amazon.com/s3/index.html[Amazon S3 Documentation] and link:https://restic.readthedocs.io/en/latest/030_preparing_a_new_repo.html#amazon-s3[Restic docs on Amazon S3].

Reviewer: I think we must say, that we support any S3 compatible servers too (open source and/or self-hosted or commertial) as well.

Author: I have added `(or S3 API compatible storage)` to the `Backup and recovery` page and the `Supported restic-compatible backup servers` page. However, I'd like to avoid mentioning `S3 API compatible storage` in the procedures because some things are likely to be different and unpredictable.

> Amazon S3:: See the documentation for link:https://docs.aws.amazon.com/s3/index.html[Amazon S3] (or the chosen S3 API compatible storage) and link:https://restic.readthedocs.io/en/latest/030_preparing_a_new_repo.html#amazon-s3[Restic docs on Amazon S3].

Reviewer 2: ```suggestion
Amazon S3:: See the documentation for link:https://docs.aws.amazon.com/s3/index.html[Amazon S3] (or the chosen S3 API compatible storage) and the link:https://restic.readthedocs.io/en/latest/030_preparing_a_new_repo.html#amazon-s3[Restic docs on Amazon S3].
```

Author: Thank you. Fixed.

> REST:: See link:https://github.com/restic/rest-server[Rest Server] and link:https://restic.readthedocs.io/en/latest/030_preparing_a_new_repo.html#rest-server[Restic docs on Rest Server].

Reviewer: I had put restic to front as it is open source, free and self-hosted solution. IMO, the easiest to set up.

Author: :) Yes, I know. I had to reconsider that based on my experience with setting up a remote server. OpenSSH is also open source and more widely used/understood. Although I'm not the intended audience here, I still found the restic server's readme page on GitHub insufficient and patchy for completing the setup, and there was no third-party website where I could find the same information explained better.

> REST:: See the README for link:https://github.com/restic/rest-server[Rest Server] and link:https://restic.readthedocs.io/en/latest/030_preparing_a_new_repo.html#rest-server[Restic docs on Rest Server].

Reviewer 2: ```suggestion
REST:: See the README for link:https://github.com/restic/rest-server[Rest Server] and the link:https://restic.readthedocs.io/en/latest/030_preparing_a_new_repo.html#rest-server[Restic docs on Rest Server].
```

Author: Thank you. Fixed.

> For testing the backing up and restoring of {prod-short} instances, {prod-short} offers the xref:backups-of-che-instances-to-the-internal-backup-server.adoc[internal backup server].

Reviewer 2: ```suggestion
To test backing up and restoring of {prod-short} instances, use the xref:backups-of-che-instances-to-the-internal-backup-server.adoc[internal backup server].
```
- Avoid giving agency to inanimate objects like software. Che can't really "offer" anything. Instead, rephrase so that human beings use the software.
